### ISSUE TITLE
Lots of queries, single in-memory doc use case

### ISSUE THREAD

**Issue by onclue** [blugelabs/bluge#78]
Hi,

Thanks a lot for this library, it looks very interesting. 
Apologies if this is not the right place for this question, but I am curious whether you think bluge can support a slightly different use case: matching lots (i.e. 1000s) of possibly complicated queries (with disjunctions and conjunctions) against a single document (5~10k words, each sentence becomes a text field). These queries basically represent something like business rules, and the task is to figure out which rules match a given document. The document doesn't need to reside in any index on disk. Would you say bluge can be a good fit for this task, or are there clear obstacles that come to mind? 
(Obviously I am not expecting a total response time of milliseconds for matching all those 1000s of queries.)

Cheers for any insights you can share!

**Comment by mschoch** [comment]
This use case is an interesting one that has seem to come up frequently recently.

My way of thinking about this is that there are sort of 2 dimensions of looking at the problem that are interesting (or at least different).

The first is the obvious one, you only have a single document vs N documents (and often people care most about large N).  A regular bluge index containing only one document will work fine, but there is a LOT of machinery to handle multiple documents that you just don't need.

Second, while you didn't call this out specifically, often the life-cycle of the "index" is different in this use-case as well.  What I mean here is that usually the whole point of building an index, is that you are going to invest some effort building a new structure, so you save time executing queries over the life of the index.  Specifically, you "get back" the time you spent building/maintaining the index in the first place.  However, in this use-case, my understanding is that often you want to run many queries against a single document right now, and then you throw that away, and do the exact same thing for a different document.  In these cases, it often is not worth building a complex index structure, because you won't run enough queries to actually save time with it.  Instead, you are better off using a simpler data-structure (often just slices so you maintain good locality).  Now much of the search just become a brute-force walk through a slice, but since we're talking about small data compactly organized, it performs well.

For the Bleve project, we created a separate index type specifically for this purpose (similar but maybe not identical to your use-case): https://github.com/blevesearch/sear  I have a prototype adaptation of this for Bluge (working name sour).  I hope to publish this soon.  The idea of both is the same, I have a single document, and I want to run searches on it (usually in a pass/fail manner)

Finally, you also mention that the queries themselves could be complicated.  I'm breaking this out separately from the other two, because I think here the impacts and solutions will be similar to a traditional Bleve/Bluge index.  What I mean is that a query with more clauses generally will take longer than one with fewer clauses.  So, the example I see often is that you have queries with clauses of the form (pseudo-syntax not real):

os:Windows/95 OR os:Windows/NT OR os:Windows/2000

Obviously this is made up, but the point is that for these cases, if you know in advance you want to find all versions of Windows, you should move this logic to the indexing phase, and index these as a single token (can even be a new synthesized field).  That way at query time you can replace these 3 clauses, with a single one looking for the "combined" token that you indexed.  In Bleve last spring, we added a Hierarchy Token Filter based on our experience working with a user that was able to use this technique.

I'm not sure if this technique will work for you, but it is general guidance we use to try and make complex queries easier to execute.

So, in summary, I expect that using our in-memory only configuration, this use-case probably works OK.  But, if you also want to throw away the "index" frequently, and repeat the process for different documents, then you'll likely be interested in the sear/sour projects I mentioned.

As I said at the beginning, this has come up several times frequently, so if you think your use-case is different, or you see a different problem with the solutions, I'd love to discuss it further.

**Comment by onclue** [comment]
Many thanks for the thoughtful answer!

You are right, there is some "inversion of control" happening in this scenario, compared to regular search engines. Here the set of queries is static and precompiled, whereas the documents are the fleeting inputs to the system, and as such there is no need to extend a document's life-cycle beyond the process of matching those queries. 

My current implementation is in fact a simple-ish algorithm where the document is indeed represented as a slice of strings (like you suggested), each of which is scanned with a big prefix trie that matches all query terms from all the rules in one pass, and then checks all relevant disjunctions and conjunctions of query terms afterwards. To make this somewhat tractable I have to keep the structure of the rules fixed, basically a list of terms that must all match (i.e. conjunction of terms to include) combined with a list of terms none of which should match (i.e. disjunction of terms to exclude) for the rule to match.  By design, any more complicated rules have to be represented by breaking it down into multiple rules of this form. This "sort of" works, and has acceptable performance as long as the matching process can be parallelized over the document's strings, but as the complexity of the rules increases I can't help but feel like I am essentially re-implementing (poorly!) what should in fact just be a proper text search engine like bluge. 

Thanks a lot for the pointers to sear/sour, I will definitely check them out and run some experiments. I will also take a closer look at your Hierarchy Token Filter suggestion, but I am not sure it applies to my use case (possibly I just misunderstood...). 

Cheers!

**Comment by mschoch** [comment]
I just published a first version of [sour](https://github.com/blugelabs/sour).  Basically, if your data is in the form of a `bluge.Document` and your queries are in the form of a `bluge.Query`, then this lets you see if they are a match.  Some attempt is made to reuse slices every time you `Reset()` to the next document.

**Comment by onclue** [comment]
That was quick, thanks a lot. Will definitely try it out.
